cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs

```diff
@@ -0,0 +1,2458 @@
+        /// <param name="attribute">The attribute to add.</param>
+        /// <param name="solutionUniqueName">The unique name of the unmanaged solution to add the attribute to.</param>
+        /// <returns>The Id of the attribute and solution component</returns>
+        public static CreateAttributeResponse CreateAttribute(this CDSWebApiService svc,
+            string entityLogicalName,
+            AttributeMetadata attribute,
+            string solutionUniqueName = null)
+        {
+            Uri uri;
+            Guid attributeMetadataId;
+            Guid solutionId;
+            Guid solutionComponentId;
+
+            if (!string.IsNullOrEmpty(solutionUniqueName))
+            {
+                //Verify that solution exists and is unmanaged.
+                try
+                {
+                    solutionId = svc.CheckSolution(solutionUniqueName);
+                }
+                catch (Exception)
+                {
+                    throw;
+                }
+            }
+
+            //Create the attribute
+            try
+            {
+                uri = svc.PostCreate($"EntityDefinitions(LogicalName='{entityLogicalName}')/Attributes", JObject.FromObject(attribute));
+            }
+            catch (Exception ex)
+            {
+                throw new Exception($"Error creating the attribute:{ex.Message}");
+            }
+
+            //Get the id from the URI
+            string id = uri.ToString().Substring(uri.ToString().IndexOf("Attributes(") + 11, 36);
+            attributeMetadataId = new Guid(id);
+
+            try
+            {
+                solutionComponentId = svc.AddSolutionComponent(attributeMetadataId, SolutionComponentType.Attribute, solutionUniqueName, true);
+            }
+            catch (Exception)
+            {
+                throw;
+            }
+
+            return new CreateAttributeResponse()
+            {
+                AttributeId = attributeMetadataId,
+                SolutionComponentId = solutionComponentId
+            };
+        }
+
+        /// <summary>
+        /// Creates a customer relationship
+        /// </summary>
+        /// <param name="svc">An instance of the CDSWebApiService to extend.</param>
+        /// <param name="lookup">The customer lookup to create.</param>
+        /// <param name="oneToManyRelationships">A pair of relationships to create for the account and contact entities.</param>
+        /// <param name="solutionUniqueName">The unique name of the unmanaged solution.</param>
+        /// <returns>Id values of the created attribute and relationships</returns>
+        public static CreateCustomerRelationshipsResponse CreateCustomerRelationships(this CDSWebApiService svc,
+            ComplexLookupAttributeMetadata lookup,
+            List<ComplexOneToManyRelationshipMetadata> oneToManyRelationships,
+            string solutionUniqueName = null)
+        {
+            if (!string.IsNullOrEmpty(solutionUniqueName))
```

Comment:
You are checking if the solution name is null or empty several times in this method. Best to check once at the beginning of the method, and also call CheckSolution(). Once that is done you are good to go for the rest of the method.

Reply:
This operation is to create a customer relationship and solutionUniqueName is an optional parameter which, if included, will associate the 3 created solution components with the unmanaged solution indicated by the solutionUniqueName.

- First check is to ensure that the solution exists and is managed.
- Second check is to only include the value in the JObject body when it exists
- Final check is after the 3 solution components are created, then they can be associated with the named solution if it exists.

In other words, the association with a solution is a secondary, optional operation within the flow. It doesn't control the main flow, just supplements it when provided. Therefore, it seems appropriate to use the existence of the parameter when needed (3 times) to control the secondary optimal options.